### PR TITLE
rocAL - Add pytorch training example

### PR DESCRIPTION
**Pull request by SundarRajan98**
- Adding pytorch training example to rocAL with options for torch, rocAL-CPU and rocAL-GPU dataloaders
- Training example adapted from https://github.com/pytorch/examples/tree/main/imagenet

### CODE REVIEW

File: rocAL_pybind/examples/rocAL_training_example/README.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
+# ImageNet training in PyTorch
+
+This implements training of popular model architectures, such as ResNet, AlexNet, and VGG on the ImageNet dataset.
+This version has been modified to use rocAL. It assumes that the dataset is raw JPEGs from the ImageNet dataset. If offers CPU and GPU based pipeline for rocAL - use rocal-cpu switch to enable CPU one and use rocal-gpu switch to enable GPU one.
```
Comment (LakshmiKumar23) [inline rocAL_pybind/examples/rocAL_training_example/README.md:4]: Please remove "one" and add `rocal-cpu` and `rocal-gpu` as a code

File: rocAL_pybind/examples/rocAL_training_example/README.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+To run use the following command
+```bash
```
Comment (LakshmiKumar23) [inline rocAL_pybind/examples/rocAL_training_example/README.md:7]: Mention this "to run distributed pytorch training"

File: rocAL_pybind/examples/rocAL_training_example/README.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+rm *.pth.tar # Remove older checkpoints saved in the folder if the example has been run before
+python3 main.py -a resnet50 --dist-url='tcp://127.0.0.1:4321' --dist-backend 'nccl' --multiprocessing-distributed --world-size 1 --rank 0 -j$(nproc) --batch-size 1024 --rocal-cpu --epochs 91 /media/imageNetCompleteDataset/
+```
+
+## Requirements
```
Comment (LakshmiKumar23) [inline rocAL_pybind/examples/rocAL_training_example/README.md:12]: requirements should come before the the above instructions to run

File: rocAL_pybind/examples/rocAL_training_example/README.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+- Install PyTorch for [ROCm](https://rocm.docs.amd.com/projects/install-on-linux/en/latest/how-to/3rd-party/pytorch-install.html)
+- Install rocAL for running rocAL trainings
+- Download the ImageNet dataset from http://www.image-net.org/
+  - Then, move and extract the training and validation images to labeled subfolders, using [the following shell script](extract_ILSVRC.sh)
+
+## Training
```
Comment (LakshmiKumar23) [inline rocAL_pybind/examples/rocAL_training_example/README.md:19]: seems like all this is coming from pytorch repo. We shouldn't be replicating line by line. Please give links

File: rocAL_pybind/examples/rocAL_training_example/main.py [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+
+    if args.dist_url == "env://" and args.world_size == -1:
+        args.world_size = int(os.environ["WORLD_SIZE"])
+
+    args.distributed = args.world_size > 1 or args.multiprocessing_distributed
+
+    if torch.cuda.is_available():
+        ngpus_per_node = torch.cuda.device_count()
+        if ngpus_per_node == 1 and args.dist_backend == "nccl":
+            warnings.warn("nccl backend >=2.5 requires GPU count>1, see https://github.com/NVIDIA/nccl/issues/103 perhaps use 'gloo'")
```
Comment (LakshmiKumar23) [inline rocAL_pybind/examples/rocAL_training_example/main.py:222]: please do not mention competitors

**Review by LakshmiKumar23 — CHANGES_REQUESTED** [review]

File: rocAL_pybind/examples/rocAL_training_example/README.md (hunk 3 above)
Comment (SundarRajan98) [reply rocAL_pybind/examples/rocAL_training_example/README.md:6]: Moved the rocAL training command after Requirements section

File: rocAL_pybind/examples/rocAL_training_example/README.md (hunk 2 above)
Comment (SundarRajan98) [reply rocAL_pybind/examples/rocAL_training_example/README.md]: The shared command is for running a training in single node, multi GPU setup. The distributed training commands are shared below in a separate Multiple nodes section

File: rocAL_pybind/examples/rocAL_training_example/README.md (hunk 4 above)
Comment (SundarRajan98) [reply rocAL_pybind/examples/rocAL_training_example/README.md:12]: Removed requirements.txt and extract_ILSVRC.sh files and updated readme to point to pytorch examples repo

File: rocAL_pybind/examples/rocAL_training_example/main.py (hunk 5 above)
Comment (SundarRajan98) [reply rocAL_pybind/examples/rocAL_training_example/main.py]: Removed the rccl repo link

File: rocAL_pybind/examples/rocAL_training_example/README.md [hunk 6]
```diff
@@ -1,20 +1,13 @@
...
-This implements training of popular model architectures, such as ResNet, AlexNet, and VGG on the ImageNet dataset.
-This version has been modified to use rocAL. It assumes that the dataset is raw JPEGs from the ImageNet dataset. If offers CPU and GPU based pipeline for rocAL - use rocal-cpu switch to enable CPU one and use rocal-gpu switch to enable GPU one.
-
-To run use the following command
-```bash
-rm *.pth.tar # Remove older checkpoints saved in the folder if the example has been run before
-python3 main.py -a resnet50 --dist-url='tcp://127.0.0.1:4321' --dist-backend 'nccl' --multiprocessing-distributed --world-size 1 --rank 0 -j$(nproc) --batch-size 1024 --rocal-cpu --epochs 91 /media/imageNetCompleteDataset/
-```
+This example implements training of popular model architectures, such as ResNet, AlexNet, and VGG on the ImageNet dataset.
+This version has been modified to use rocAL. It assumes that the dataset is raw JPEGs from the ImageNet dataset. If offers CPU and GPU based pipeline for rocAL - use `rocal-cpu` switch to enable CPU and use `rocal-gpu` switch to enable GPU.
```
Comment (LakshmiKumar23) [inline rocAL_pybind/examples/rocAL_training_example/README.md:4]: please use the term 'argument' instead of switch. We pass it to the command right?

**Comment by LakshmiKumar23** [comment]
@SundarRajan98 please move this folder under [examples/pytorch](https://github.com/ROCm/rocAL/tree/develop/docs/examples/pytorch). Rename it to `imagenet_training` subfolder and add both the files under there. Change the filename to 'imagenet_training.py`

File: rocAL_pybind/examples/rocAL_training_example/main.py [hunk 7]
```diff
@@ -219,7 +219,7 @@ def main():
     if torch.cuda.is_available():
         ngpus_per_node = torch.cuda.device_count()
         if ngpus_per_node == 1 and args.dist_backend == "nccl":
-            warnings.warn("nccl backend >=2.5 requires GPU count>1, see https://github.com/NVIDIA/nccl/issues/103 perhaps use 'gloo'")
+            warnings.warn("nccl backend >=2.5 requires GPU count>1, perhaps use 'gloo'")
```
[the remark this answers is not in the document]
Comment (SundarRajan28) [reply rocAL_pybind/examples/rocAL_training_example/main.py]: Used autopep8 to format the entire training script

**Comment by SundarRajan28** [comment]
@LakshmiKumar23 Renamed and moved the train example folder to docs/examples/pytorch

**Comment by LakshmiKumar23** [comment]
@SundarRajan28 Have you noticed this issue? Running on this docker `swethabs/rocm6.0.2_pytorch_2.1.2_mivisionx_py3.10` on Lockhart
```
root@x1000c4s5b1n0:/dockerx/rocAL/docs/examples/pytorch/imagenet_training# python3 imagenet_training.py -a resnet50 -j$(nproc) --batch-size 1024 --rocal-gpu /media/imagenet_20dir/
=> creating model 'resnet50'
OK: OpenVX using GPU device - 0: AMD Instinct MI250X [gfx90a:sramecc+:xnack-] with 110 CUs on PCI bus c1:00.0

OK: loaded 114 kernels from libvx_rpp.so
Pipeline has been created succesfully
rocal "gpu" variant
OK: OpenVX using GPU device - 0: AMD Instinct MI250X [gfx90a:sramecc+:xnack-] with 110 CUs on PCI bus c1:00.0

OK: loaded 114 kernels from libvx_rpp.so
Pipeline has been created succesfully
rocal "gpu" variant
OK: HIP buffer usage: 7247794256, 6/6
Bus error (core dumped)
root@x1000c4s5b1n0:/dockerx/rocAL/docs/examples/pytorch/imagenet_training# python3 imagenet_training.py -a resnet50 -j$(nproc) --batch-size 1024 --rocal-cpu /media/imagenet_20dir/
=> creating model 'resnet50'
OK: loaded 114 kernels from libvx_rpp.so
Pipeline has been created succesfully
rocal "cpu" variant
OK: loaded 114 kernels from libvx_rpp.so
Pipeline has been created succesfully
rocal "cpu" variant
Bus error (core dumped)
```

**Comment by SundarRajan28** [comment]
@LakshmiKumar23 Please try in x1000c4s5b0n0 node. The x1000c4s5b1n0 node has issues with trainings. Also please try the command in single node, multiple GPUs section of the readme. The above command runs training in CPU.

**Review by LakshmiKumar23 — APPROVED** [review]
Tested with resnet50 example on a single node, multiple GPU. @kiritigowda can be merged

**Pull request merged**